[CT] Integer constant-time selection without CMOV on i386

hbrodin opened the pull request:
Implement constant time logic for integers on non-CMOV capable targets.
- Only promote i8 to i32 when CMOV available, use native 8-bit otherwise
- Add post-RA expansion for CTSELECT_I386_GR* pseudos with temp registers

hbrodin committed:
[CT] Integer constant-time selection without CMOV on i386

Implement constant time logic for integers on non-CMOV capable targets.
- Only promote i8 to i32 when CMOV available, use native 8-bit otherwise
- Add post-RA expansion for CTSELECT_I386_GR* pseudos with temp registers

hbrodin commented:
Currently, invoking ctselect on `i64` values fails as `EFLAGS` get overwritten by the first 32-bit expansion of the type-legalized `i64` and causes the second 32-bit value to be incorrect. Working on it.

hbrodin commented:
The issue is resolved now.

kumarak commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -25083,7 +25083,8 @@ SDValue X86TargetLowering::LowerCTSELECT(SDValue Op, SelectionDAG &DAG) const {
   // Promote small integer types to avoid partial register stalls
-  if ((Op.getValueType() == MVT::i8) ||
+  // Exception: For i8 without CMOV, prefer native 8-bit constant-time operations
+  if ((Op.getValueType() == MVT::i8 && Subtarget.canUseCMOV()) ||
```
Is there a specific reason to add checks for CMOV here? I didn't get comment about native constant time operation for `i8`. Is it not going to fallback to `CTSELECT` anyway?

kumarak replied:
Resolving it since the change is there is reduce the instructions generated by `ISD::ANY_EXTEND`

kumarak commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -37560,122 +37561,69 @@ X86TargetLowering::emitPatchableEventCall(MachineInstr &MI,
+  // Step 1: Materialize condition byte from EFLAGS
+  // This is done OUTSIDE the constant-time bundle, before any EFLAGS corruption
+  Register CondByteReg = MRI.createVirtualRegister(&X86::GR8RegClass);
+  BuildMI(*BB, MI, MIMD, TII->get(X86::SETCCr), CondByteReg).addImm(OppCC);
```
Do you see eflag corruption because of some other instruction getting generated between `TEST` and `SETcc` affecting flags? It is nice to find what is the actual reason.

hbrodin replied:
So, the `testb` happened before the bundle, then in the bundle some operations also set `EFLAGS`, then in bundle number 2, a new `sete` was done but now it operated on a clobbered `EFLAGS`. 
By doing it in this new way, the generated code for 64-bit cases results in a single `sete` which is shared between both bundles (type legalized 64->2x32bit). Makes sense?

kumarak replied:
It makes sense. We don't want to set the bytes based on condition for both lower and higher order 32 bits.

kumarak commented on llvm/lib/Target/X86/X86InstrCMovSetCC.td:
```diff
@@ -114,15 +114,15 @@ let Uses = [EFLAGS], isNotDuplicable = 1, isPseudo = 1 in {
   multiclass CTSELECT<X86TypeInfo t> {
     // register-only
-    let isCommutable = 0, SchedRW = [WriteCMOV],
+    let isCommutable = 0, SchedRW = [WriteCMOV], Predicates = [HasCMOV],
```
Do we need to have `HasCMOV` predicate here? Is this to avoid conflict with `CTSELECT_I386` variant?

kumarak replied:
I see it was there with `defm CTSELECT16` definition and you moved it here.

kumarak commented:
I tested the changes with unit tests and it looks good to. We can merge it after resolving comments.

hbrodin committed:
[CT] Make cmov predicates more stringent

Previously, in some cases, multiple patterns could match depending
on what attributes where used. Now we are solely relying on the hasCMOV.

Clarified a comment about why we are not widening 8-bit values in non-cmov
context.

kumarak commented:
Merging it now since it looks good to me. We can have other changes for float/vector types in subsequent PRs.
